### PR TITLE
Still throw a locked exception when the path is not relative to $user/files/

### LINKED ISSUE #5219: Updating NC12: update theming app fails
Same issue than the one reported  [here](https://help.nextcloud.com/t/solved-updating-nc12-update-theming-app-fails/12930) , but currently unsolved for me.
Giving NC a try, but I got the same issue while migrating/updating.

I migrated from OC 9.1 to NC 11.0.3.2 and then updating to NC12 using occ:
```
# sudo -u www-data php occ upgrade
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Set log level to debug
Updating database schema
Updated database
Updating <theming> ...
InvalidArgumentException: $absolutePath must be relative to "files"
Update failed
Maintenance mode is kept active
Reset log level
```

System:
debian 8.8, x86_64
PHP 5.6.30
NGINX 1.10.3
mariadb 10.0.30

Nextcloud logs:
   
```
{"reqId":"JoG4zF47JsDfNLinxANy","level":0,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"core","method":"--","url":"--","message":"starting upgrade from 11.0.3.2 to 12.0.0.29","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Repair::step: Repair step: Repair MySQL collation","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Repair::info: Repair info: All tables already have the correct collation -> nothing to do","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Repair::step: Repair step: Repair SQLite autoincrement","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Repair::step: Repair step: Copy data from accounts table when migrating from ownCloud","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:01+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::dbUpgradeBefore: Updating database schema","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:02+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::dbUpgrade: Updated database","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::appUpgradeStarted: Updating <theming> ...","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Repair::step: Repair step: Move theming files to AppData storage","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":3,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"$absolutePath must be relative to \"files\", value is \"%s\"","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":3,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"core","method":"--","url":"--","message":"Exception: {\"Excepti
    on\":\"InvalidArgumentException\",\"Message\":\"$absolutePath must be relative to \\\"files\\\"\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(1951): OC\\\\Files\\\\View->getPathRelativeToFiles('\\\/appdata_oc9nyo...')\\n#1 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(1126): OC\\\\Files\\\\View->changeLock('\\\/appdata_oc9nyo...', 2)\\n#2 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(269): OC\\\\Files\\\\View->basicOperation('mkdir', '\\\/appdata_oc9nyo...', Array)\\n#3 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Node\\\/Folder.php(160): OC\\\\Files\\\\View->mkdir('\\\/appdata_oc9nyo...')\\n#4 [internal function]: OC\\\\Files\\\\Node\\\\Folder->newFolder('appdata_oc9nyok...')\\n#5 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Node\\\/LazyRoot.php(63): call_user_func_array(Array, Array)\\n#6 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Node\\\/LazyRoot.php(385): OC\\\\Files\\\\Node\\\\LazyRoot->__call('newFolder', Array)\\n#7 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/AppData\\\/AppData.php(82): OC\\\\Files\\\\Node\\\\LazyRoot->newFolder('appdata_oc9nyok...')\\n#8 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/AppData\\\/AppData.php(112): OC\\\\Files\\\\AppData\\\\AppData->getAppDataFolder()\\n#9 \\\/home\\\/www-data\\\/nextcloud\\\/apps\\\/theming\\\/lib\\\/Migration\\\/ThemingImages.php(56): OC\\\\Files\\\\AppData\\\\AppData->newFolder('images')\\n#10 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Repair.php(85): OCA\\\\Theming\\\\Migration\\\\ThemingImages->run(Object(OC\\\\Repair))\\n#11 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/legacy\\\/app.php(1114): OC\\\\Repair->run()\\n#12 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/legacy\\\/app.php(1049): OC_App::executeRepairSteps('theming', Array)\\n#13 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Updater.php(381): OC_App::updateApp('theming')\\n#14 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Updater.php(264): OC\\\\Updater->doAppUpgrade()\\n#15 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Updater.php(130): OC\\\\Updater->doUpgrade('12.0.0.29', '11.0.3.2')\\n#16 \\\/home\\\/www-data\\\/nextcloud\\\/core\\\/Command\\\/Upgrade.php(263): OC\\\\Updater->upgrade()\\n#17 \\\/home\\\/www-data\\\/nextcloud\\\/3rdparty\\\/symfony\\\/console\\\/Command\\\/Command.php(256): OC\\\\Core\\\\Command\\\\Upgrade->execute(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#18 \\\/home\\\/www-data\\\/nextcloud\\\/3rdparty\\\/symfony\\\/console\\\/Application.php(818): Symfony\\\\Component\\\\Console\\\\Command\\\\Command->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#19 \\\/home\\\/www-data\\\/nextcloud\\\/3rdparty\\\/symfony\\\/console\\\/Application.php(186): Symfony\\\\Component\\\\Console\\\\Application->doRunCommand(Object(OC\\\\Core\\\\Command\\\\Upgrade), Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#20 \\\/home\\\/www-data\\\/nextcloud\\\/3rdparty\\\/symfony\\\/console\\\/Application.php(117): Symfony\\\\Component\\\\Console\\\\Application->doRun(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#21 \\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Console\\\/Application.php(170): Symfony\\\\Component\\\\Console\\\\Application->run(NULL, NULL)\\n#22 \\\/home\\\/www-data\\\/nextcloud\\\/console.php(92): OC\\\\Console\\\\Application->run()\\n#23 \\\/home\\\/www-data\\\/nextcloud\\\/occ(11): require_once('\\\/home\\\/www-data\\\/...')\\n#24 {main}\",\"File\":\"\\\/home\\\/www-data\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php\",\"Line\":2084}","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":3,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::failure: InvalidArgumentException: $absolutePath must be relative to \"files\"","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":3,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::updateEnd: Update failed","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::maintenanceActive: Maintenance mode is kept active","userAgent":"--","version":"11.0.3.2"}
    {"reqId":"JoG4zF47JsDfNLinxANy","level":1,"time":"2017-05-31T17:23:03+02:00","remoteAddr":"","user":"--","app":"updater","method":"--","url":"--","message":"\\OC\\Updater::resetLogLevel: Reset log level to Error(3)","userAgent":"--","version":"11.0.3.2"}


```
And since` $absolutePath must be relative to \"files\", value is \"%s\"` does not help very much ;-), it seems that $absolutePath is set to "remoteAddr"

```
# sudo -u www-data php occ app:list
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Enabled:
  - activity: 2.5.2
  - comments: 1.2.0
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - federation: 1.2.0
  - files: 1.7.2
  - files_external: 1.3.0
  - files_pdfviewer: 1.1.1
  - files_sharing: 1.4.0
  - files_texteditor: 2.4.1
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - files_videoplayer: 1.1.0
  - firstrunwizard: 2.1
  - gallery: 17.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: true
  - password_policy: 1.2.2
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharebymail: 1.2.0
  - survey_client: 1.0.0
  - systemtags: 1.2.0
  - theming: 1.1.1
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - workflowengine: 1.2.0
Disabled:
  - admin_audit
  - encryption
  - user_external
  - user_ldap

```

```
# sudo -u www-data php occ config:list system
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
          ***REMOVED SENSITIVE VALUE***
        ],
        "datadirectory": "\/var\/www\/nextcloud_data\/data",
        "overwrite.cli.url": "\/owncloud",
        "dbtype": "mysql",
        "version": "11.0.3.2",
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "loglevel": 3,
        "logtimezone": "Europe\/Paris",
        "theme": "",
        "maintenance": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "trashbin_retention_obligation": "auto",
        "appstore.experimental.enabled": true
    }
}
```

**Are you using encryption:** no

**Are you using an external user-backend, no

### PR DESCRIPTION
**Pull request by nickvergessen**
This turns `InvalidArgumentException: $absolutePath must be relative to "files"` from #5219 into a locked exception. Now the question is, why is the `appdata_` folder locked?

@LukasReschke is it okay to throw with the path, if it is not relative to a users `files/` directory?

Fix #5219

### CODE REVIEW

File: lib/private/Files/View.php [hunk 1]
```diff
@@ -2050,7 +2057,7 @@ protected function shouldLockFile($path) {
 			return ($pathSegments[2] === 'files') && (count($pathSegments) > 3);
 		}
 
-		return true;
+		return strpos($path, '/appdata_') !== 0;
```
Comment (nickvergessen) [inline lib/private/Files/View.php]: @rullzer we locked everything that was only 1 level deep, which includes `/appdata_<id>`

So this should be fixed now...

Comment (nickvergessen) [inline lib/private/Files/View.php]: Or should we even make this always false?

**Review by LukasReschke — APPROVED** [review]
Reported to fix this issue multiple times in our forum.

> @LukasReschke is it okay to throw with the path, if it is not relative to a users files/ directory?

Works for me.

File: lib/private/Files/View.php (hunk 1 above)
Comment (douzepouze) [inline lib/private/Files/View.php]: Fixed the issure for me.

**Comment by icewind1991** [comment]
This doesn't actually lock files within appdata though only the appdata folder itself.

**Comment by rullzer** [comment]
Any reason not to just lock all files? Instead of only select parts?

**Comment by icewind1991** [comment]
Performance mostly.

Locking `/$user/files` is useless since nothing does a destructive operation on it that would require an exclusive lock

**Review by MorrisJobke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
@nickvergessen Backport to stable12?

**Comment by MorrisJobke** [comment]
The CI job in here was fine, but seems that the unit tests got updated in the meantime. Now they fail: https://drone.nextcloud.com/nextcloud/server/1185/71/sqlite-php7.0-samba-non-native

@nickvergessen Could you have a look at this?

**Pull request merged**